openshift-kube-apiserver/openshiftkubeapiserver/patch_handlerchain.go

```diff
@@ -95,3 +105,104 @@ func translateLegacyScopeImpersonation(handler http.Handler) http.Handler {
 		handler.ServeHTTP(w, req)
 	})
 }
+
+// withWatchRateLimit rate limits the establishment of watches to keep kube-apiservers from crashing.
+// Rate limiting watches effectively creates an upper bound on secret and configmap mounts of 10*QPS because of a
+// ten minute watch timeout and kubelets use watches to get the content for the mount.
+// We will break the rate limiting into three timespans
+// 1. first ten minutes: this is the most restrictive 10000 total mounted secrets and configmaps, 1000 per minute.
+//    We're trying to break up the slug of kubelet traffic and
+//    we want to be sure that operators can make progress during this time if we need to recover a cluster in
+//    a bad state.
+// 2. second ten minutes: this is less restrictive  20000 total mounted secrets and configmaps, 2000 per minute.
+//    This lets us start to ramp up during a relative steady state.
+// 3. no limit.  We have this to handle cases of large clusters with more than 20000 mounted secrets and configmaps.
+//    I honestly don't know how common this is, but I don't want to break on it.
+// Recall that we observed more than the the 30,000 per minute observed during some disruptive events on a cluster.
+// In addition, we special case watches in the platform operator namespaces, cluster scope, and kube-system.
+// We have not observed large numbers of these and they are required in order to make progress when trying to correct
+// some kinds of cluster failures.
+func withWatchRateLimit(handler http.Handler) http.Handler {
+	earlyRateLimiter := flowcontrol.NewTokenBucketRateLimiter(16.6, 100)
```

Comment:
how did we get this number?

Reply:
> how did we get this number?

1000 per minute.  it's 30x less than the peak from an unhappy cluster.

Reply:
16.6*60 = 1000, per minute => 10000 per 10 min